README.md
<div align="center">

# dashed_line

[![pub.dev][pub-badge]][pub-link] [![build status][build-badge]][build-link]

</div>

![][underheader]

Draw dashed lines with any shape and style you want. Just like that.

## Usage

Add the dependency to your `pubspec.yaml` (you can see the newest version in the pub badge):

```yaml
dependencies:
  dashed_line: ^0.1.0
```

Use the widget:

```dart
import 'package:dashed_line/dashed_line.dart';

// ...

DashedLine(
  path: Path()..cubicTo(-40, 53, 14, 86, 61, 102),
  color: Colors.red,
)
```

...you can also use the SVG path commands instead of a `Path`:

```dart
DashedLine.svgPath(
  'C -40 53 14 86 61 102',
  color: Colors.red,
)
```

## Where to obtain path methods/commands?

There are many ways.

You can construct the path yourself using the [`Path` methods][path-methods] like [`lineTo`][path-lineto] or [`cubicTo`][path-cubicto].

You can also use the [path commands][svg-commands] used in the `d` attribute of SVG files. You can do this manually or via export from one of the vector graphics software, like [Inkscape][inkscape] or [Figma][figma]. We found [SvgPathEditor][svgpatheditor] quite useful too.

### Example: Exporting path commands from Figma

|               First step              |             Second step            |
|:-------------------------------------:|:----------------------------------:|
| ![Create path using pen][figma-step1] | ![Export path as SVG][figma-step2] |

And now we have an SVG in the clipboard:

```svg
<svg width="190" height="48" viewBox="0 0 190 48" fill="none" xmlns="http://www.w3.org/2000/svg">
  <path d="M1 16C73 79 100 3 121 0.999997C142 -1 214.5 11.5 179 47" stroke="black"/>
</svg>
```

From here we can just copy the value of the `path`'s `d` attribute and use for the `DashedLine.svgPath` constructor's first argument.

## 🚨 Note about paths

Due to how `Path`s work in Flutter and Skia, the `DashedLine` widget takes NOT as much space as the dashed line needs, but as much it needs to contain all the [control points][control-points]. Table below should help understand the problem.

| Path commands                 |          Path          |          Result line          |
|-------------------------------|:----------------------:|:-----------------------------:|
| ``` C 8 63 14 86 61 102 ```   | ![][path-inside-bbox]  | ![][path-inside-bbox-result]  |
| ``` C -40 53 14 86 61 102 ``` | ![][path-outside-bbox] | ![][path-outside-bbox-result] |

## License

See [`LICENSE`][license].

[underheader]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/underheader.png
[pub-badge]: https://img.shields.io/pub/v/dashed_line
[pub-link]: https://pub.dev/packages/dashed_line
[build-badge]: https://img.shields.io/github/workflow/status/leancodepl/dashed_line/Test
[build-link]: https://github.com/leancodepl/dashed_line/actions/workflows/test.yml

[path-methods]: https://api.flutter.dev/flutter/dart-ui/Path-class.html
[path-lineto]: https://api.flutter.dev/flutter/dart-ui/Path/lineTo.html
[path-cubicto]: https://api.flutter.dev/flutter/dart-ui/Path/cubicTo.html
[svg-commands]: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d
[inkscape]: https://inkscape.org/
[figma]: https://www.figma.com/
[svgpatheditor]: https://yqnn.github.io/svg-path-editor/

[figma-step1]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/figma-step1.png
[figma-step2]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/figma-step2.png

[control-points]: https://en.wikipedia.org/wiki/Control_point_(mathematics)
[path-inside-bbox]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/path-inside-bbox.png
[path-inside-bbox-result]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/path-inside-bbox-result.png
[path-outside-bbox]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/path-outside-bbox.png
[path-outside-bbox-result]: https://raw.githubusercontent.com/leancodepl/dashed_line/main/art/path-outside-bbox-result.png

[license]: LICENSE


CHANGELOG.md
# 0.1.0+2

- Fix build badge in README.

# 0.1.0+1

- Fix README image paths.

# 0.1.0

- First release.


test/dashed_line_test.dart
import 'package:dashed_line/dashed_line.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const black = Color(0xFF000000);
  const red = Color(0xFFFF0000);

  group('Path', () {
    final path = Path()
      ..moveTo(1, 52)
      ..cubicTo(67, 128, 125, 11, 155, 22)
      ..cubicTo(189, 34, 172, 75, 143, 76)
      ..cubicTo(114, 77, 102, 36, 123, 16)
      ..cubicTo(156, -16, 235, 11, 255, 29);

    testWidgets('Default arguments', (tester) async {
      final line = DashedLine(path: path, color: black);
      await testDashedLine(tester, line, 'path-default');
    });

    testWidgets('Line fit = fill', (tester) async {
      final line = DashedLine(path: path, color: black, lineFit: LineFit.fill);
      await testDashedLine(tester, line, 'path-fill');
    });

    testWidgets('All custom arguments', (tester) async {
      final line = DashedLine(
        path: path,
        color: red,
        dashLength: 30,
        dashSpace: 10,
        dashCap: StrokeCap.round,
        width: 5,
      );

      await testDashedLine(tester, line, 'path-custom');
    });
  });

  group('SVG path', () {
    const svgPath =
        'M0.999939 51.5989C66.9999 127.599 124.966 10.9987 155 21.5989C189 '
        '33.5989 172 74.5989 143 75.5989C114 76.5989 102 35.5989 123 '
        '15.5989C156 -15.8296 235 10.5989 255 28.5989';

    testWidgets('Default arguments', (tester) async {
      final line = DashedLine.svgPath(svgPath, color: black);
      await testDashedLine(tester, line, 'svg-default');
    });

    testWidgets('All custom arguments', (tester) async {
      final line = DashedLine.svgPath(
        svgPath,
        color: red,
        dashLength: 30,
        dashSpace: 10,
        dashCap: StrokeCap.round,
        width: 5,
      );

      await testDashedLine(tester, line, 'svg-custom');
    });
  });
}

Future<void> testDashedLine(
  WidgetTester tester,
  DashedLine dashedLine,
  String goldenFile,
) async {
  await tester.pumpWidget(
    Container(
      color: const Color(0xFFFFFFFF),
      alignment: Alignment.center,
      child: Container(
        width: 200,
        height: 200,
        decoration: BoxDecoration(border: Border.all()),
        child: dashedLine,
      ),
    ),
  );
  await expectLater(
    find.byType(DashedLine),
    matchesGoldenFile('goldens/$goldenFile.png'),
  );
}


lib/src/dashed_line.dart
import 'package:dashed_line/dashed_line.dart';
import 'package:dashed_line/src/fitting_path_size.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:svg_path_parser/svg_path_parser.dart';

import 'dashed_line_painter.dart';

/// A dashed line following a path.
///
/// The line follows a specified path, either:
/// 1. A [Path] if using a default constructor.
/// 2. An SVG path definition when using the [DashedLine.svgPath] constructor.
///
/// The `lineFit`, `alignment`, `dashLength`, `dashSpace`, `strokeCap`,
/// and `strokeWidth` are customizable.
class DashedLine extends StatelessWidget {
  /// Creates a dashed line following a `path`.
  const DashedLine({
    Key? key,
    required this.path,
    required this.color,
    this.lineFit = LineFit.contain,
    this.alignment = Alignment.center,
    this.dashLength = 4,
    this.dashSpace = 8,
    this.dashCap = StrokeCap.butt,
    this.width = 1,
  }) : super(key: key);

  /// Creates a dashed line following a path defined using SVG path commands.
  ///
  /// The `svgPath` is a path definition, i.e. a list of path commands
  /// known from the SVG file format.
  ///
  /// See also:
  /// - https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d#path_commands
  DashedLine.svgPath(
    String svgPath, {
    Key? key,
    required this.color,
    this.lineFit = LineFit.contain,
    this.alignment = Alignment.center,
    this.dashLength = 4,
    this.dashSpace = 8,
    this.dashCap = StrokeCap.butt,
    this.width = 1,
  })  : path = parseSvgPath(svgPath),
        super(key: key);

  /// The path that the dashed line follows.
  ///
  /// Can consist of multiple path metrics.
  final Path path;

  /// The dashed line color.
  final Color color;

  /// The [LineFit] of the line inside the parent.
  final LineFit lineFit;

  /// The alignment of the dashed line relative to the parent.
  final Alignment alignment;

  /// The length of a single dash, i.e. the drawn parts of the whole [path].
  final double dashLength;

  /// The distance between each dash, i.e. the blank spaces on the whole [path].
  final double dashSpace;

  /// The [StrokeCap] of each dash.
  ///
  /// This is the shape that appears on the beginning and end of each dash
  /// on the dashed line.
  final StrokeCap dashCap;

  /// The width of the dashed line.
  final double width;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        final size = fittingPathSize(path, constraints.biggest, lineFit);

        return CustomPaint(
          painter: DashedLinePainter(
            path: path,
            color: color,
            lineFit: lineFit,
            alignment: alignment,
            dashLength: dashLength,
            dashSpace: dashSpace,
            strokeCap: dashCap,
            strokeWidth: width,
          ),
          child: SizedBox.fromSize(size: size),
        );
      },
    );
  }
}


lib/src/fitting_path_size.dart
import 'dart:math' show min;
import 'dart:ui';

import 'line_fit.dart';

/// Calculates the scale by which the [path] should be scaled to fit in a
/// container of size [containerSize], taking [lineFit] into account.
///
/// The [Offset] returned should be interpreted as a scale in X and Y axes.
Offset fittingPathScale(Path path, Size containerSize, LineFit lineFit) {
  final pathSize = path.getBounds().size;

  if (lineFit == LineFit.contain) {
    final scale = min(
      containerSize.width / pathSize.width,
      containerSize.height / pathSize.height,
    );

    return Offset(scale, scale);
  } else {
    return Offset(
      containerSize.width / pathSize.width,
      containerSize.height / pathSize.height,
    );
  }
}

/// Calculates the size required to accomodate the [path] in a container of size
/// [containerSize], taking [lineFit] into account.
Size fittingPathSize(Path path, Size containerSize, LineFit lineFit) {
  final pathSize = path.getBounds().size;
  final scale = fittingPathScale(path, containerSize, lineFit);

  return Size(pathSize.width * scale.dx, pathSize.height * scale.dy);
}


lib/src/dashed_line_painter.dart
import 'dart:typed_data';

import 'package:flutter/rendering.dart';

import 'fitting_path_size.dart';
import 'line_fit.dart';

class DashedLinePainter extends CustomPainter {
  DashedLinePainter({
    required this.path,
    required this.color,
    required this.lineFit,
    required this.alignment,
    required this.dashLength,
    required this.dashSpace,
    required this.strokeCap,
    required this.strokeWidth,
  });

  final Path path;

  /// The color of the dashed line.
  final Color color;

  final LineFit lineFit;
  final Alignment alignment;
  final double dashLength;
  final double dashSpace;

  /// The width of the edges, given in logical pixels measured in the direction
  /// orthogonal to the direction of the path.
  final double strokeWidth;

  /// The kind of finish to place on the end of line.
  final StrokeCap strokeCap;

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color
      ..strokeCap = strokeCap
      ..strokeWidth = strokeWidth
      ..style = PaintingStyle.stroke;

    final scale = fittingPathScale(path, size, lineFit);

    final scaledPathSize = fittingPathSize(path, size, lineFit);
    // Offset required to align the dashed line accordingly to the alignment.
    final alignmentOffset = Offset(
      (size.width - scaledPathSize.width) / 2 * (alignment.x + 1),
      (size.height - scaledPathSize.height) / 2 * (alignment.y + 1),
    );

    final pathBounds = path.getBounds();
    // Offset needed to move the path into the painter bounds so
    // that it doesn't overflow.
    final negativeOffset = Offset(
      -pathBounds.left * scale.dx,
      -pathBounds.top * scale.dy,
    );

    final pathMetrics = path.computeMetrics();
    for (final pathMetric in pathMetrics) {
      for (var dist = 0.0;
          dist < pathMetric.length;
          dist += dashLength + dashSpace) {
        final dashPath = pathMetric
            .extractPath(dist, dist + dashLength)
            .transform(_scaleMatrix4(scale))
            .shift(alignmentOffset + negativeOffset);

        canvas.drawPath(dashPath, paint);
      }
    }
  }

  /// https://en.wikipedia.org/wiki/Scaling_(geometry)#Using_homogeneous_coordinates
  static Float64List _scaleMatrix4(Offset scale) {
    final matrix = <List<double>>[
      [scale.dx, 0, 0, 0],
      [0, scale.dy, 0, 0],
      [0, 0, 1, 0],
      [0, 0, 0, 1],
    ];

    return Float64List.fromList(matrix.expand((x) => x).toList());
  }

  @override
  bool shouldRepaint(DashedLinePainter oldDelegate) => path != oldDelegate.path;
}


example/lib/main.dart
import 'package:dashed_line/dashed_line.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: HomeScreen()));

class HomeScreen extends StatelessWidget {
  const HomeScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final path = Path()..cubicTo(-40, 53, 14, 86, 61, 102);

    return Scaffold(
      appBar: AppBar(
        title: const Text('dashed_line example'),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {},
        child: const Icon(Icons.celebration),
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          const Spacer(),
          Text(
            'Click me!',
            textAlign: TextAlign.center,
            style: Theme.of(context).textTheme.headline4,
          ),
          Expanded(
            child: Padding(
              padding: const EdgeInsets.fromLTRB(72, 0, 72, 64),
              child: DashedLine(
                path: path,
                color: Theme.of(context).textTheme.headline4!.color!,
              ),
            ),
          ),
        ],
      ),
    );
  }
}
